File: app/src/main/java/com/quran/labs/androidquran/ui/util/TypefaceManager.java
package com.quran.labs.androidquran.ui.util;

import android.content.Context;
import android.graphics.Typeface;
import androidx.annotation.NonNull;
import com.quran.labs.androidquran.data.QuranFileConstants;

public class TypefaceManager {
  public static final int TYPE_UTHMANI_HAFS = 1;
  public static final int TYPE_NOOR_HAYAH = 2;
  public static final int TYPE_UTHMANIC_WARSH = 3;
  public static final int TYPE_UTHMANIC_QALOON = 4;

  private static Typeface typeface;
  private static Typeface arabicTafseerTypeface;
  private static Typeface arabicHeaderFooterTypeface;

  public static Typeface getUthmaniTypeface(@NonNull Context context) {
    if (typeface == null) {
      final String fontName;
      switch (QuranFileConstants.FONT_TYPE) {
        case TYPE_NOOR_HAYAH: {
          fontName = "noorehira.ttf";
          break;
        }
        case TYPE_UTHMANIC_WARSH: {
          fontName = "uthmanic_warsh_ver09.ttf";
          break;
        }
        case TYPE_UTHMANIC_QALOON: {
          fontName = "uthmanic_qaloon_ver21.ttf";
          break;
        }
        case TYPE_UTHMANI_HAFS:
        default: {
          fontName = "uthmanic_hafs_ver12.otf";
        }
      }
      typeface = Typeface.createFromAsset(context.getAssets(), fontName);
    }
    return typeface;
  }

  public static Typeface getTafseerTypeface(@NonNull Context context) {
    if (arabicTafseerTypeface == null) {
      arabicTafseerTypeface = Typeface.createFromAsset(context.getAssets(), "kitab.ttf");
    }
    return arabicTafseerTypeface;
  }

  public static Typeface getHeaderFooterTypeface(@NonNull Context context) {
    if (arabicHeaderFooterTypeface == null) {
      arabicHeaderFooterTypeface = Typeface.createFromAsset(context.getAssets(), "UthmanTN1Ver10.otf");
    }
    return arabicHeaderFooterTypeface;
  }
}

File: app/src/main/java/com/quran/labs/androidquran/ui/util/TypefaceManager.kt
package com.quran.labs.androidquran.ui.util

import android.content.Context
import android.graphics.Typeface
import com.quran.labs.androidquran.data.QuranFileConstants

object TypefaceManager {

  const val TYPE_UTHMANI_HAFS = 1
  const val TYPE_NOOR_HAYAH = 2
  const val TYPE_UTHMANIC_WARSH = 3
  const val TYPE_UTHMANIC_QALOON = 4

  private var typeface: Typeface? = null
  private var arabicTafseerTypeface: Typeface? = null
  private var arabicHeaderFooterTypeface: Typeface? = null
  private var dyslexicTypeface: Typeface? = null

  @JvmStatic
  fun getUthmaniTypeface(context: Context): Typeface {
    return typeface ?: run {
      val fontName = when (QuranFileConstants.FONT_TYPE) {
        TYPE_NOOR_HAYAH -> "noorehira.ttf"
        TYPE_UTHMANIC_WARSH -> "uthmanic_warsh_ver09.ttf"
        TYPE_UTHMANIC_QALOON -> "uthmanic_qaloon_ver21.ttf"
        else -> "uthmanic_hafs_ver12.otf"
      }
      val instance = Typeface.createFromAsset(context.assets, fontName)
      typeface = instance
      instance
    }
  }

  fun getTafseerTypeface(context: Context): Typeface {
    return arabicTafseerTypeface ?: run {
      val instance = Typeface.createFromAsset(context.assets, "kitab.ttf")
      arabicTafseerTypeface = instance
      instance
    }
  }

  fun getDyslexicTypeface(context: Context): Typeface {
    return dyslexicTypeface ?: run {
      val instance = Typeface.createFromAsset(context.assets, "OpenDyslexic.otf")
      dyslexicTypeface = instance
      instance
    }
  }

  fun getHeaderFooterTypeface(context: Context): Typeface {
    return arabicHeaderFooterTypeface ?: run {
      val instance = Typeface.createFromAsset(context.assets, "UthmanTN1Ver10.otf")
      arabicHeaderFooterTypeface = instance
      instance
    }
  }

}
